cmd/tealdbg/debugger_test.go

```diff
@@ -185,3 +191,91 @@ func TestSession(t *testing.T) {
 	require.NotEmpty(t, name)
 	require.Greater(t, len(data), 0)
 }
+
+// Tests control functions for stepping over subroutines and checks
+// that call stack is inspected correctly.
+func TestCallStackControl(t *testing.T) {
+	partitiontest.PartitionTest(t)
+	s := createSessionFromSource(t, "#pragma version %d\nlab1:\nint 1\ncallsub lab1\ndup\n+\n")
+	s.callStack = []logic.CallFrame{
+		{FrameLine: 2, LabelName: "lab1"},
+	}
+	// Check for step over on a callsub line
+	s.line.Store(3)
+
+	err := s.SetBreakpoint(3)
+	require.NoError(t, err)
+
+	ackCount := 0
+	initialStackDepth := len(s.callStack)
+	done := make(chan struct{})
+
+	// Update function.
+	ackFunc := func() {
+		ackCount++
+		<-s.acknowledged
+		done <- struct{}{}
+	}
+
+	go ackFunc()
+
+	s.StepOver()
+	<-done
+
+	require.Equal(t, 4, s.debugConfig.BreakAtLine)
+	require.Equal(t, breakpoint{true, true}, s.breakpoints[4])
+	require.Equal(t, true, s.debugConfig.StepOutOver)
+
+	require.Equal(t, 1, ackCount)
+	require.Equal(t, initialStackDepth, len(s.callStack))
+
+	// Check for step over on a non callsub line
+	s.line.Store(4)
+
+	go ackFunc()
+
+	s.StepOver()
+	<-done
+
+	require.Equal(t, true, s.debugConfig.StepBreak)
+	require.Equal(t, false, s.debugConfig.StepOutOver)
+
+	require.Equal(t, 2, ackCount)
+	require.Equal(t, initialStackDepth, len(s.callStack))
+
+	// Check that step out when call stack depth is 1 sets breakpoint to the
```

Comment:
`s.line` actually never changes in any of these tests (`s.line` is the state where the evaluator has currently paused in the program). We create a fake "ackFunc" instead of actually updating the line and callstack (which is called by the evaluator in the `eval.go` package). 

For control like `stepOver`, we actually inspect the current line to make sure it has a `callsub` to determine how we will set the breakpoint, but the other control functions do not, so the `s.line` value won't matter in the unit tests. Let me know if this answers your question.

Reply:
@algochoi Thanks - That addresses the question.  Though it opens other questions in my mind.  😅 

I'll resolve here and hold for a reply on https://github.com/algorand/go-algorand/pull/3653#discussion_r834577692 while continuing review.  I think the linked thread will  cover some of what I have in mind.